File: packages/omdaa-python/src/omdaa/resources/resources.py
# All API resources - mirror omdaa-js SDK paths and methods.
from urllib.parse import quote


class MessagesResource:
    def __init__(self, client):
        self._client = client

    def send_text(self, params):
        body = {
            "sessionId": params.get("sessionId"),
            "to": params.get("to"),
            "message": params.get("message") or params.get("text"),
            "quotedMessageId": params.get("quotedMessageId"),
            "quoted": params.get("quoted"),
        }
        return self._client.post("/messages/send-text", body)

    def send_image(self, params):
        return self._client.post("/messages/send-image", params)

    def get_received(self, params=None):
        return self._client.get("/messages/received", params=params)

    def get_sent(self, params=None):
        return self._client.get("/messages/sent", params=params)

    def get_stats(self, params=None):
        return self._client.get("/messages/stats", params=params)

    def get_by_chat(self, remote_jid, params=None):
        return self._client.get(f"/messages/chat/{quote(remote_jid)}", params=params)

    def search(self, params=None):
        return self._client.get("/messages/search", params=params)

    def get_recent_chats(self, params=None):
        return self._client.get("/messages/recent-chats", params=params)

    def get_chats(self, params=None):
        return self._client.get("/messages/chats", params=params)

    def delete_message(self, message_id):
        return self._client.delete(f"/messages/{quote(message_id)}")

    def forward(self, body):
        return self._client.post("/messages/forward", body)

    def forward_multiple(self, body):
        return self._client.post("/messages/forward-multiple", body)

    def delete_multiple(self, body):
        return self._client.delete("/messages/multiple", body)

    def schedule(self, body):
        return self._client.post("/messages/schedule", body)

    def send_location(self, body):
        return self._client.post("/messages/send-location", body)

    def send_contact(self, body):
        return self._client.post("/messages/send-contact", body)

    def send_voice(self, body):
        return self._client.post("/messages/send-voice", body)

    def send_audio(self, body):
        return self._client.post("/messages/send-audio", body)

    def send_reaction(self, body):
        return self._client.post("/messages/reaction", body)

    def remove_reaction(self, body):
        return self._client.delete("/messages/reaction", body)

    def send_read_receipt(self, body):
        return self._client.post("/messages/read-receipt", body)

    def mark_chat_read(self, body):
        return self._client.post("/messages/mark-chat-read", body)

    def get_read_status(self, message_id):
        return self._client.get(f"/messages/read-status/{quote(message_id)}")

    def update_read_receipts_privacy(self, body):
        return self._client.put("/messages/privacy/read-receipts", body)

    def send_typing(self, body):
        return self._client.post("/messages/typing", body)

    def send_recording(self, body):
        return self._client.post("/messages/recording", body)

    def send_file(self, body):
        return self._client.post("/messages/send-file", body)

    def send_media(self, body):
        return self._client.post("/messages/send-media", body)

    def send_poll(self, body):
        return self._client.post("/messages/send-poll", body)

    def send_list(self, body):
        return self._client.post("/messages/send-list", body)

    def send_buttons(self, body):
        return self._client.post("/messages/send-buttons", body)

    def send_status(self, body):
        return self._client.post("/messages/send-status", body)

    def send_media_advanced(self, body):
        return self._client.post("/messages/send-media-advanced", body)

    def send_ptv(self, body):
        return self._client.post("/messages/send-ptv", body)

    def send_sticker(self, body):
        return self._client.post("/messages/send-sticker", body)

    def get_poll_results(self, poll_id, params=None):
        return self._client.get(f"/messages/poll/{quote(poll_id)}/results", params=params)


class SessionsResource:
    def __init__(self, client):
        self._client = client

    def list(self):
        return self._client.get("/sessions")

    def create(self, params=None):
        return self._client.post("/sessions/create", params or {})

    def get(self, session_id):
        return self._client.get(f"/sessions/{quote(session_id)}")

    def get_qr(self, session_id):
        return self._client.get(f"/sessions/{quote(session_id)}/qr")

    def disconnect(self, session_id):
        return self._client.delete(f"/sessions/{quote(session_id)}/logout")

    def delete(self, session_id):
        return self._client.delete(f"/sessions/delete/{quote(session_id)}")


class WebhooksResource:
    def __init__(self, client):
        self._client = client

    def get(self):
        return self._client.get("/webhooks")

    def set(self, params):
        return self._client.post("/webhooks/set", params)

    def remove(self):
        return self._client.delete("/webhooks/remove")

    def test(self):
        return self._client.get("/webhooks/test")

    def get_events(self, params=None):
        return self._client.get("/webhooks/events", params=params)

    def get_stats(self):
        return self._client.get("/webhooks/stats")


class TemplatesResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/templates", params=params)

    def get(self, template_id):
        return self._client.get(f"/templates/{quote(template_id)}")

    def create(self, body):
        return self._client.post("/templates", body)

    def delete(self, template_id):
        return self._client.delete(f"/templates/{quote(template_id)}")


class ScheduledResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/scheduled", params=params)

    def get(self, scheduled_id):
        return self._client.get(f"/scheduled/{quote(scheduled_id)}")

    def cancel(self, scheduled_id):
        return self._client.delete(f"/scheduled/{quote(scheduled_id)}")


class BulkResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/bulk", params=params)

    def get(self, bulk_id):
        return self._client.get(f"/bulk/{quote(bulk_id)}")

    def create(self, body):
        return self._client.post("/bulk", body)

    def cancel(self, bulk_id):
        return self._client.delete(f"/bulk/{quote(bulk_id)}")


class ContactsResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/contacts", params=params)

    def get(self, contact_id):
        return self._client.get(f"/contacts/{quote(contact_id)}")

    def create(self, body):
        return self._client.post("/contacts", body)

    def update(self, contact_id, body):
        return self._client.put(f"/contacts/{quote(contact_id)}", body)

    def delete(self, contact_id):
        return self._client.delete(f"/contacts/{quote(contact_id)}")


class GroupsResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/groups", params=params)

    def get(self, group_id):
        return self._client.get(f"/groups/{quote(group_id)}")

    def create(self, body):
        return self._client.post("/groups", body)

    def update(self, group_id, body):
        return self._client.put(f"/groups/{quote(group_id)}", body)

    def leave(self, group_id):
        return self._client.post(f"/groups/{quote(group_id)}/leave", {})


class ChatsResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/chats", params=params)

    def get(self, chat_id):
        return self._client.get(f"/chats/{quote(chat_id)}")

    def archive(self, chat_id, body=None):
        return self._client.post(f"/chats/{quote(chat_id)}/archive", body or {})

    def unarchive(self, chat_id):
        return self._client.post(f"/chats/{quote(chat_id)}/unarchive", {})


class BusinessResource:
    def __init__(self, client):
        self._client = client

    def get(self, params=None):
        return self._client.get("/business", params=params)

    def update(self, body):
        return self._client.put("/business", body)


class StorageResource:
    def __init__(self, client):
        self._client = client

    def get_usage(self, params=None):
        return self._client.get("/storage/usage", params=params)

    def list_files(self, params=None):
        return self._client.get("/storage/files", params=params)


class LabelsResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/labels", params=params)

    def get(self, label_id):
        return self._client.get(f"/labels/{quote(label_id)}")

    def create(self, body):
        return self._client.post("/labels", body)

    def delete(self, label_id):
        return self._client.delete(f"/labels/{quote(label_id)}")


class IntegrationsResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/integrations", params=params)

    def get(self, integration_id):
        return self._client.get(f"/integrations/{quote(integration_id)}")

    def connect(self, body):
        return self._client.post("/integrations/connect", body)

    def disconnect(self, integration_id):
        return self._client.delete(f"/integrations/{quote(integration_id)}")


class QueueResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/queue", params=params)

    def get(self, job_id):
        return self._client.get(f"/queue/{quote(job_id)}")

    def cancel(self, job_id):
        return self._client.delete(f"/queue/{quote(job_id)}")


class InteractiveResource:
    def __init__(self, client):
        self._client = client

    def send_list(self, body):
        return self._client.post("/interactive/list", body)

    def send_buttons(self, body):
        return self._client.post("/interactive/buttons", body)


class ApiKeysResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/api-keys", params=params)

    def create(self, body):
        return self._client.post("/api-keys", body)

    def revoke(self, key_id):
        return self._client.delete(f"/api-keys/{quote(key_id)}")


class ProfileResource:
    def __init__(self, client):
        self._client = client

    def get(self):
        return self._client.get("/profile")

    def update(self, body):
        return self._client.put("/profile", body)


class DashboardResource:
    def __init__(self, client):
        self._client = client

    def get(self, params=None):
        return self._client.get("/dashboard", params=params)


class PlansResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/plans", params=params)

    def get(self, plan_id):
        return self._client.get(f"/plans/{quote(plan_id)}")


class BillingResource:
    def __init__(self, client):
        self._client = client

    def get(self):
        return self._client.get("/billing")

    def get_invoices(self, params=None):
        return self._client.get("/billing/invoices", params=params)


class ReferralResource:
    def __init__(self, client):
        self._client = client

    def get(self):
        return self._client.get("/referral")

    def get_links(self, params=None):
        return self._client.get("/referral/links", params=params)


class NotificationsResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/notifications", params=params)

    def mark_read(self, notification_id):
        return self._client.post(f"/notifications/{quote(notification_id)}/read", {})


class MetricsResource:
    def __init__(self, client):
        self._client = client

    def get(self, params=None):
        return self._client.get("/metrics", params=params)


class BackupResource:
    def __init__(self, client):
        self._client = client

    def create(self, body=None):
        return self._client.post("/backup", body or {})

    def list(self, params=None):
        return self._client.get("/backup", params=params)

    def restore(self, backup_id):
        return self._client.post(f"/backup/{quote(backup_id)}/restore", {})


class SecurityResource:
    def __init__(self, client):
        self._client = client

    def get_settings(self):
        return self._client.get("/security/settings")

    def update_settings(self, body):
        return self._client.put("/security/settings", body)


class AuditResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/audit", params=params)

    def get(self, audit_id):
        return self._client.get(f"/audit/{quote(audit_id)}")


class OrganizationsResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/organizations", params=params)

    def get(self, org_id):
        return self._client.get(f"/organizations/{quote(org_id)}")

    def create(self, body):
        return self._client.post("/organizations", body)

    def update(self, org_id, body):
        return self._client.put(f"/organizations/{quote(org_id)}", body)


class SettingsResource:
    def __init__(self, client):
        self._client = client

    def get(self):
        return self._client.get("/settings")

    def update(self, body):
        return self._client.put("/settings", body)


class ChannelsResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/channels", params=params)

    def get(self, channel_id):
        return self._client.get(f"/channels/{quote(channel_id)}")


class PrivacyResource:
    def __init__(self, client):
        self._client = client

    def get(self):
        return self._client.get("/privacy")

    def update(self, body):
        return self._client.put("/privacy", body)


class SupportResource:
    def __init__(self, client):
        self._client = client

    def tickets(self, params=None):
        return self._client.get("/support/tickets", params=params)

    def create_ticket(self, body):
        return self._client.post("/support/tickets", body)


class ProxyResource:
    def __init__(self, client):
        self._client = client

    def get(self):
        return self._client.get("/proxy")

    def update(self, body):
        return self._client.put("/proxy", body)


class CallsResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/calls", params=params)

    def get(self, call_id):
        return self._client.get(f"/calls/{quote(call_id)}")


class AuthResource:
    def __init__(self, client):
        self._client = client

    def login(self, body):
        return self._client.post("/auth/login", body)

    def logout(self):
        return self._client.post("/auth/logout", {})

    def refresh(self, body=None):
        return self._client.post("/auth/refresh", body or {})

    def me(self):
        return self._client.get("/auth/me")


class UsersResource:
    def __init__(self, client):
        self._client = client

    def list(self, params=None):
        return self._client.get("/users", params=params)

    def get(self, user_id):
        return self._client.get(f"/users/{quote(user_id)}")

    def create(self, body):
        return self._client.post("/users", body)

    def update(self, user_id, body):
        return self._client.put(f"/users/{quote(user_id)}", body)

    def delete(self, user_id):
        return self._client.delete(f"/users/{quote(user_id)}")


class EmailResource:
    def __init__(self, client):
        self._client = client

    def send(self, body):
        return self._client.post("/email/send", body)

    def get_settings(self):
        return self._client.get("/email/settings")


class AiResource:
    def __init__(self, client):
        self._client = client

    def chat(self, body):
        return self._client.post("/ai/chat", body)

    def completions(self, body):
        return self._client.post("/ai/completions", body)

File: packages/omdaa-go/resources.go
package omdaa

import "net/url"

// Messages
type MessagesResource struct{ c *Client }

func (r *MessagesResource) SendText(params map[string]interface{}) (map[string]interface{}, error) {
	body := map[string]interface{}{
		"sessionId": params["sessionId"],
		"to":        params["to"],
		"message":   params["message"],
	}
	if params["message"] == nil {
		body["message"] = params["text"]
	}
	if params["quotedMessageId"] != nil {
		body["quotedMessageId"] = params["quotedMessageId"]
	}
	if params["quoted"] != nil {
		body["quoted"] = params["quoted"]
	}
	return r.c.Post("/messages/send-text", body)
}
func (r *MessagesResource) SendImage(params map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/messages/send-image", params)
}
func (r *MessagesResource) GetReceived(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/messages/received", params)
}
func (r *MessagesResource) GetSent(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/messages/sent", params)
}
func (r *MessagesResource) GetStats(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/messages/stats", params)
}
func (r *MessagesResource) GetByChat(remoteJid string, params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/messages/chat/"+url.PathEscape(remoteJid), params)
}
func (r *MessagesResource) Search(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/messages/search", params)
}
func (r *MessagesResource) DeleteMessage(messageId string) (map[string]interface{}, error) {
	return r.c.Delete("/messages/"+url.PathEscape(messageId), nil)
}
func (r *MessagesResource) SendFile(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/messages/send-file", body)
}
func (r *MessagesResource) SendReaction(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/messages/reaction", body)
}

// Sessions
type SessionsResource struct{ c *Client }

func (r *SessionsResource) List() (map[string]interface{}, error) {
	return r.c.Get("/sessions", nil)
}
func (r *SessionsResource) Create(params map[string]interface{}) (map[string]interface{}, error) {
	if params == nil {
		params = map[string]interface{}{}
	}
	return r.c.Post("/sessions/create", params)
}
func (r *SessionsResource) Get(sessionId string) (map[string]interface{}, error) {
	return r.c.Get("/sessions/"+url.PathEscape(sessionId), nil)
}
func (r *SessionsResource) GetQR(sessionId string) (map[string]interface{}, error) {
	return r.c.Get("/sessions/"+url.PathEscape(sessionId)+"/qr", nil)
}
func (r *SessionsResource) Disconnect(sessionId string) (map[string]interface{}, error) {
	return r.c.Delete("/sessions/"+url.PathEscape(sessionId)+"/logout", nil)
}
func (r *SessionsResource) Delete(sessionId string) (map[string]interface{}, error) {
	return r.c.Delete("/sessions/delete/"+url.PathEscape(sessionId), nil)
}

// Webhooks
type WebhooksResource struct{ c *Client }

func (r *WebhooksResource) Get() (map[string]interface{}, error) {
	return r.c.Get("/webhooks", nil)
}
func (r *WebhooksResource) Set(params map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/webhooks/set", params)
}
func (r *WebhooksResource) Remove() (map[string]interface{}, error) {
	return r.c.Delete("/webhooks/remove", nil)
}
func (r *WebhooksResource) Test() (map[string]interface{}, error) {
	return r.c.Get("/webhooks/test", nil)
}

// Templates
type TemplatesResource struct{ c *Client }

func (r *TemplatesResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/templates", params)
}
func (r *TemplatesResource) Get(templateId string) (map[string]interface{}, error) {
	return r.c.Get("/templates/"+url.PathEscape(templateId), nil)
}
func (r *TemplatesResource) Create(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/templates", body)
}
func (r *TemplatesResource) Delete(templateId string) (map[string]interface{}, error) {
	return r.c.Delete("/templates/"+url.PathEscape(templateId), nil)
}

// Scheduled, Bulk, Contacts, Groups, Chats
type ScheduledResource struct{ c *Client }

func (r *ScheduledResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/scheduled", params)
}
func (r *ScheduledResource) Get(id string) (map[string]interface{}, error) {
	return r.c.Get("/scheduled/"+url.PathEscape(id), nil)
}
func (r *ScheduledResource) Cancel(id string) (map[string]interface{}, error) {
	return r.c.Delete("/scheduled/"+url.PathEscape(id), nil)
}

type BulkResource struct{ c *Client }

func (r *BulkResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/bulk", params)
}
func (r *BulkResource) Get(id string) (map[string]interface{}, error) {
	return r.c.Get("/bulk/"+url.PathEscape(id), nil)
}
func (r *BulkResource) Create(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/bulk", body)
}

type ContactsResource struct{ c *Client }

func (r *ContactsResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/contacts", params)
}
func (r *ContactsResource) Get(id string) (map[string]interface{}, error) {
	return r.c.Get("/contacts/"+url.PathEscape(id), nil)
}
func (r *ContactsResource) Create(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/contacts", body)
}

type GroupsResource struct{ c *Client }

func (r *GroupsResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/groups", params)
}
func (r *GroupsResource) Get(id string) (map[string]interface{}, error) {
	return r.c.Get("/groups/"+url.PathEscape(id), nil)
}
func (r *GroupsResource) Create(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/groups", body)
}

type ChatsResource struct{ c *Client }

func (r *ChatsResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/chats", params)
}
func (r *ChatsResource) Get(id string) (map[string]interface{}, error) {
	return r.c.Get("/chats/"+url.PathEscape(id), nil)
}

// Business, Storage, Labels, Integrations, Queue, Interactive
type BusinessResource struct{ c *Client }

func (r *BusinessResource) Get(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/business", params)
}
func (r *BusinessResource) Update(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Put("/business", body)
}

type StorageResource struct{ c *Client }

func (r *StorageResource) GetUsage(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/storage/usage", params)
}

type LabelsResource struct{ c *Client }

func (r *LabelsResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/labels", params)
}
func (r *LabelsResource) Create(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/labels", body)
}
func (r *LabelsResource) Delete(id string) (map[string]interface{}, error) {
	return r.c.Delete("/labels/"+url.PathEscape(id), nil)
}

type IntegrationsResource struct{ c *Client }

func (r *IntegrationsResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/integrations", params)
}
func (r *IntegrationsResource) Get(id string) (map[string]interface{}, error) {
	return r.c.Get("/integrations/"+url.PathEscape(id), nil)
}

type QueueResource struct{ c *Client }

func (r *QueueResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/queue", params)
}
func (r *QueueResource) Get(id string) (map[string]interface{}, error) {
	return r.c.Get("/queue/"+url.PathEscape(id), nil)
}
func (r *QueueResource) Cancel(id string) (map[string]interface{}, error) {
	return r.c.Delete("/queue/"+url.PathEscape(id), nil)
}

type InteractiveResource struct{ c *Client }

func (r *InteractiveResource) SendList(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/interactive/list", body)
}
func (r *InteractiveResource) SendButtons(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/interactive/buttons", body)
}

// ApiKeys, Profile, Dashboard, Plans, Billing, Referral
type ApiKeysResource struct{ c *Client }

func (r *ApiKeysResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/api-keys", params)
}
func (r *ApiKeysResource) Create(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/api-keys", body)
}
func (r *ApiKeysResource) Revoke(id string) (map[string]interface{}, error) {
	return r.c.Delete("/api-keys/"+url.PathEscape(id), nil)
}

type ProfileResource struct{ c *Client }

func (r *ProfileResource) Get() (map[string]interface{}, error) {
	return r.c.Get("/profile", nil)
}
func (r *ProfileResource) Update(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Put("/profile", body)
}

type DashboardResource struct{ c *Client }

func (r *DashboardResource) Get(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/dashboard", params)
}

type PlansResource struct{ c *Client }

func (r *PlansResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/plans", params)
}
func (r *PlansResource) Get(id string) (map[string]interface{}, error) {
	return r.c.Get("/plans/"+url.PathEscape(id), nil)
}

type BillingResource struct{ c *Client }

func (r *BillingResource) Get() (map[string]interface{}, error) {
	return r.c.Get("/billing", nil)
}
func (r *BillingResource) GetInvoices(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/billing/invoices", params)
}

type ReferralResource struct{ c *Client }

func (r *ReferralResource) Get() (map[string]interface{}, error) {
	return r.c.Get("/referral", nil)
}
func (r *ReferralResource) GetLinks(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/referral/links", params)
}

// Notifications, Metrics, Backup, Security, Audit
type NotificationsResource struct{ c *Client }

func (r *NotificationsResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/notifications", params)
}
func (r *NotificationsResource) MarkRead(id string) (map[string]interface{}, error) {
	return r.c.Post("/notifications/"+url.PathEscape(id)+"/read", map[string]interface{}{})
}

type MetricsResource struct{ c *Client }

func (r *MetricsResource) Get(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/metrics", params)
}

type BackupResource struct{ c *Client }

func (r *BackupResource) Create(body map[string]interface{}) (map[string]interface{}, error) {
	if body == nil {
		body = map[string]interface{}{}
	}
	return r.c.Post("/backup", body)
}
func (r *BackupResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/backup", params)
}
func (r *BackupResource) Restore(id string) (map[string]interface{}, error) {
	return r.c.Post("/backup/"+url.PathEscape(id)+"/restore", map[string]interface{}{})
}

type SecurityResource struct{ c *Client }

func (r *SecurityResource) GetSettings() (map[string]interface{}, error) {
	return r.c.Get("/security/settings", nil)
}
func (r *SecurityResource) UpdateSettings(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Put("/security/settings", body)
}

type AuditResource struct{ c *Client }

func (r *AuditResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/audit", params)
}
func (r *AuditResource) Get(id string) (map[string]interface{}, error) {
	return r.c.Get("/audit/"+url.PathEscape(id), nil)
}

// Organizations, Settings, Channels, Privacy, Support, Proxy, Calls
type OrganizationsResource struct{ c *Client }

func (r *OrganizationsResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/organizations", params)
}
func (r *OrganizationsResource) Get(id string) (map[string]interface{}, error) {
	return r.c.Get("/organizations/"+url.PathEscape(id), nil)
}
func (r *OrganizationsResource) Create(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/organizations", body)
}
func (r *OrganizationsResource) Update(id string, body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Put("/organizations/"+url.PathEscape(id), body)
}

type SettingsResource struct{ c *Client }

func (r *SettingsResource) Get() (map[string]interface{}, error) {
	return r.c.Get("/settings", nil)
}
func (r *SettingsResource) Update(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Put("/settings", body)
}

type ChannelsResource struct{ c *Client }

func (r *ChannelsResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/channels", params)
}
func (r *ChannelsResource) Get(id string) (map[string]interface{}, error) {
	return r.c.Get("/channels/"+url.PathEscape(id), nil)
}

type PrivacyResource struct{ c *Client }

func (r *PrivacyResource) Get() (map[string]interface{}, error) {
	return r.c.Get("/privacy", nil)
}
func (r *PrivacyResource) Update(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Put("/privacy", body)
}

type SupportResource struct{ c *Client }

func (r *SupportResource) Tickets(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/support/tickets", params)
}
func (r *SupportResource) CreateTicket(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/support/tickets", body)
}

type ProxyResource struct{ c *Client }

func (r *ProxyResource) Get() (map[string]interface{}, error) {
	return r.c.Get("/proxy", nil)
}
func (r *ProxyResource) Update(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Put("/proxy", body)
}

type CallsResource struct{ c *Client }

func (r *CallsResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/calls", params)
}
func (r *CallsResource) Get(id string) (map[string]interface{}, error) {
	return r.c.Get("/calls/"+url.PathEscape(id), nil)
}

// Auth, Users, Email, Ai
type AuthResource struct{ c *Client }

func (r *AuthResource) Login(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/auth/login", body)
}
func (r *AuthResource) Logout() (map[string]interface{}, error) {
	return r.c.Post("/auth/logout", map[string]interface{}{})
}
func (r *AuthResource) Refresh(body map[string]interface{}) (map[string]interface{}, error) {
	if body == nil {
		body = map[string]interface{}{}
	}
	return r.c.Post("/auth/refresh", body)
}
func (r *AuthResource) Me() (map[string]interface{}, error) {
	return r.c.Get("/auth/me", nil)
}

type UsersResource struct{ c *Client }

func (r *UsersResource) List(params map[string]string) (map[string]interface{}, error) {
	return r.c.Get("/users", params)
}
func (r *UsersResource) Get(id string) (map[string]interface{}, error) {
	return r.c.Get("/users/"+url.PathEscape(id), nil)
}
func (r *UsersResource) Create(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/users", body)
}
func (r *UsersResource) Update(id string, body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Put("/users/"+url.PathEscape(id), body)
}
func (r *UsersResource) Delete(id string) (map[string]interface{}, error) {
	return r.c.Delete("/users/"+url.PathEscape(id), nil)
}

type EmailResource struct{ c *Client }

func (r *EmailResource) Send(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/email/send", body)
}
func (r *EmailResource) GetSettings() (map[string]interface{}, error) {
	return r.c.Get("/email/settings", nil)
}

type AiResource struct{ c *Client }

func (r *AiResource) Chat(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/ai/chat", body)
}
func (r *AiResource) Completions(body map[string]interface{}) (map[string]interface{}, error) {
	return r.c.Post("/ai/completions", body)
}
